src/runtime_src/tools/xclbinutil/SectionIPLayout.cxx

```diff
@@ -266,60 +351,91 @@ SectionIPLayout::marshalFromJSON(const boost::property_tree::ptree& _ptSection,
     ipDataHdr.m_type = getIPType(sm_type);
 
     // For these IPs, the struct indices needs to be initialized
-    if ((ipDataHdr.m_type == IP_MEM_DDR4) ||
-        (ipDataHdr.m_type == IP_MEM_HBM) ||
-        (ipDataHdr.m_type == IP_MEM_HBM_ECC)) {
-      ipDataHdr.indices.m_index = ptIPData.get<uint16_t>("m_index");
-      ipDataHdr.indices.m_pc_index = ptIPData.get<uint8_t>("m_pc_index", 0);
-    } else {
-      // Get the properties value (if one is defined)
-      auto sProperties = ptIPData.get<std::string>("properties", "0");
-      ipDataHdr.properties = (uint32_t)XUtil::stringToUInt64(sProperties);
-
-      { // m_int_enable
-        boost::optional<bool> bIntEnable;
-        bIntEnable = ptIPData.get_optional<bool>("m_int_enable");
-        if (bIntEnable.is_initialized()) {
-          ipDataHdr.properties = ipDataHdr.properties & (~(uint32_t)IP_INT_ENABLE_MASK);  // Clear existing bit
-          if (bIntEnable.get()) {
-            ipDataHdr.properties = ipDataHdr.properties | ((uint32_t)IP_INT_ENABLE_MASK); // Set bit
+    switch (ipDataHdr.m_type) {
+      case IP_MEM_DDR4:
+      case IP_MEM_HBM:
+      case IP_MEM_HBM_ECC:
+      {
+        ipDataHdr.indices.m_index = ptIPData.get<uint16_t>("m_index");
+        ipDataHdr.indices.m_pc_index = ptIPData.get<uint8_t>("m_pc_index", 0);
+        break;
+      }
+      case IP_PS_KERNEL:
+      {
+        { // m_subtype
+          auto sSubType = ptIPData.get<std::string>("m_subtype", "");
+          if (!sSubType.empty()) {
+            ipDataHdr.ps_kernel.m_subtype = (unsigned int)getSubType(sSubType);
           }
         }
-      }
-
-      { // m_interrupt_id
-        auto sInterruptID = ptIPData.get<std::string>("m_interrupt_id", "");
-        if (!sInterruptID.empty()) {
-          unsigned int interruptID = std::stoul(sInterruptID);
-          unsigned int maxValue = ((unsigned int)IP_INTERRUPT_ID_MASK) >> IP_INTERRUPT_ID_SHIFT;
-          if (interruptID > maxValue) {
-            auto errMsg = boost::format("ERROR: The m_interrupt_id (%d), exceeds maximum value (%d).") % interruptID % maxValue;
-            throw std::runtime_error(errMsg.str());
+  
+        { // m_functinoal
+          auto sFunctional = ptIPData.get<std::string>("m_functional", "");
+          if (!sFunctional.empty()) {
```

Comment:
I like how you did this.  This is much better then using size() for size() in many cases could be a calculated value, while empty fast and simple.